multiraft/heartbeat_test.go

```diff
@@ -0,0 +1,228 @@
+// for names of contributors.
+//
+// Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
+
+package multiraft
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+// processEventsUntil reads and acknowledges messages from the given channel
+// until either the given conditional returns true, the channel is closed or a
+// read on the channel times out.
+func processEventsUntil(ch <-chan *interceptMessage, f func(*RaftMessageRequest) bool) {
+	t := time.After(500 * time.Millisecond)
+	for {
+		select {
+		case e, ok := <-ch:
+			if !ok {
+				return
+			}
+			e.ack <- struct{}{}
+			if f(e.args.(*RaftMessageRequest)) {
+				return
+			}
+		case <-t:
+			log.Error("timeout when reading from intercept channel")
+			return
+		}
+	}
+}
+
+// a heartbeatCondition is invoked when determining whether the intercepted
+// stream should be let go. The heartbeatCountMap reflects the heartbeats
+// intercepted up to and including the given *RaftMessageRequest.
+type heartbeatCondition func(*RaftMessageRequest, heartbeatCountMap) bool
+
+func alwaysFalse(r *RaftMessageRequest) bool {
+	return false
+}
+
+// a heartbeatCountMap helps us keep track of the amount of heartbeats sent and
+// received for a number of nodes, indexed by their NodeID.
+type heartbeatCountMap map[uint64]heartbeatCount
+
+func (hcm heartbeatCountMap) Sum() int {
+	sum := 0
+	for _, v := range hcm {
+		sum += v.Sum()
+	}
+	return sum
+}
+
+func (hcm heartbeatCountMap) String() string {
+	max := uint64(0)
+	for k := range hcm {
+		if k > max {
+			max = k
+		}
+	}
+	ret := "{ "
+	for i := uint64(1); i <= max; i++ {
+		ret += fmt.Sprintf("%d: %s, ", i, hcm[i])
+	}
+	return ret + "}"
+}
+
+type heartbeatCount struct {
+	reqOut, reqIn, respOut, respIn int
+}
+
+func (hc heartbeatCount) Sum() int {
+	return hc.reqOut + hc.reqIn + hc.respOut + hc.respIn
+}
+
+func (hc heartbeatCount) String() string {
+	return fmt.Sprintf("[reqOut=%d reqIn=%d respOut=%d respIn=%d]",
+		hc.reqOut, hc.reqIn, hc.respOut, hc.respIn)
+}
+
+// countHeartbeats reads intercepted messages from the channel until the given
+// conditional returns true (or the channel closes or times out). The returned
+// heartbeatCountMap will contain the count of heartbeat requests and responses
+// for each nodeID observed in the message stream.
+func countHeartbeats(ch <-chan *interceptMessage,
+	cond heartbeatCondition) heartbeatCountMap {
+
+	cnt := make(heartbeatCountMap)
+	processEventsUntil(ch, func(req *RaftMessageRequest) bool {
+		from := cnt[req.Message.From]
+		to := cnt[req.Message.To]
+		switch req.Message.Type {
+		case raftpb.MsgHeartbeat:
+			from.reqOut++
+			to.reqIn++
+		case raftpb.MsgHeartbeatResp:
+			from.respOut++
+			to.respIn++
+		default:
+			// Don't evaluate cond() when this is
+			// not a heartbeat.
+			return false
+		}
+		cnt[req.Message.From] = from
```

Comment:
It's convenient though because I can avoid the usual `v, ok := ...; if !ok { v[key] = ... }` code.